packages/serverpod/lib/src/server/diagnostic_events/operation_context.dart

```diff
@@ -0,0 +1,198 @@
+import 'package:uuid/uuid_value.dart';
+
+import '../../authentication/authentication_info.dart';
+
+import 'event_handler.dart';
+import 'connection_info.dart';
+
+/// Represents the type of "operation" during which the event was caused.
+///
+/// An operation is a client-originated call / operation or a scheduled operation,
+/// i.e. a method call, a streaming call, a web call, or a future call.
+enum OperationType {
+  /// A future call.
+  future,
+
+  /// A web call.
+  web,
+
+  /// A method call.
+  method,
+
+  /// A streaming call / message.
+  stream,
+
+  /// An internal operation.
+  /// Typically corresponds to an [InternalSession].
+  internal,
+}
+
+/// Represents the operation that the event was caused by or in relation to.
+/// An operation is a client-originated call / operation or a scheduled operation,
+/// i.e. a method call, a streaming call, a web call, or a future call.
+base class OperationEventContext extends DiagnosticEventContext {
+  /// The type of operation.
+  final OperationType operationType;
+
+  /// The session ID of the operation, if available.
+  final UuidValue? sessionId;
+
+  /// The user authentication information, if available.
+  final AuthenticationInfo? userAuthInfo;
+
+  /// Creates a new [OperationEventContext].
+  const OperationEventContext({
+    required super.serverName,
+    required super.serverId,
+    required super.serverRunMode,
+    required this.operationType,
+    this.sessionId,
+    this.userAuthInfo,
+  });
+
+  @override
+  Map<String, dynamic> toJson() {
+    return {
+      ...super.toJson(),
+      'operationType': '$operationType',
+      'sessionId': sessionId != null ? '$sessionId' : null,
+      'userAuthInfo': userAuthInfo != null ? '$userAuthInfo' : null,
+    };
+  }
+
+  @override
+  String toString() {
+    return '$runtimeType($serverName, $serverId, $serverRunMode, '
+        '$operationType, $sessionId, $userAuthInfo)';
+  }
+}
+
+/// The context of a future call operation.
+base class FutureCallOpContext extends OperationEventContext {
+  /// The name of the future call.
+  final String futureCallName;
+
+  /// Creates a new [FutureCallOpContext].
+  const FutureCallOpContext({
+    required super.serverName,
+    required super.serverId,
+    required super.serverRunMode,
+    required super.sessionId,
+    super.userAuthInfo,
+    required this.futureCallName,
+  }) : super(operationType: OperationType.future);
+
+  @override
+  Map<String, dynamic> toJson() {
+    return {
+      ...super.toJson(),
+      'futureCallName': futureCallName,
+    };
+  }
+}
+
+/// The context of a client call operation.
+/// Subclasses represent web, method, and method streaming calls.
+base class ClientCallOpContext extends OperationEventContext {
```

Comment:
Question/Discussion: Do we need different context objects for each call context? To me this almost looks like what the Session object should have been.

Would it be possible to consolidate to a single class? Is that even desirable?